core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -436,6 +436,29 @@ public static void verifyTypeEquivalence(
     return mapping;
   }
 
+  /**
+   * Returns a permutation describing where the Project's fields come from
+   * after the Project is pushed down.
+   */
+  public static Mappings.TargetMapping permutationPushDownProject(
+      List<RexNode> nodes,
+      RelDataType inputRowType,
+      int sourceOffset,
+      int targetOffset) {
+    final Mappings.TargetMapping mapping =
+        Mappings.create(MappingType.PARTIAL_FUNCTION,
+            inputRowType.getFieldCount() + sourceOffset,
+            nodes.size() + targetOffset);
+    for (Ord<RexNode> node : Ord.zip(nodes)) {
+      if (node.e instanceof RexInputRef) {
```

Comment:
Should we still consider the collation shifting when the project node is not a `inputRef`, i.e. how about there is a 
monotonically increasing funtion call ?

Reply:
It seems unnecessary to do that，since RelFieldCollation is only associated with RexInputRef, and the index of the operand of RexCall which is RexInputRef has not been changed after Project was push down